### DISCUSSION TITLE
Configurate text and value for prevalues in dropdown

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#11084] [Features and ideas]
In property editors like dropdown, checkbox list and radiobutton list we can configurate prevalues.

![image](https://user-images.githubusercontent.com/2919859/133050863-880ba77f-553a-414b-995f-39cca517cf0d.png)

However currently there is no way to configurate both text and value and also default selected option(s).
Often you don't want the label is the same as the stored value, but currently we may need to resolve the value in e.g. razor view, so `None` would be mapped to a value like `g-0` (no gutter).

It would could specify the specific value (which editors don't see), it would in many cases require less code in views.

Not sure what the best approach would be, but it could be something like this with toggles. By default it could generate a lowercase value from the label/text and replace e.g. space with hyphen.

![image](https://user-images.githubusercontent.com/2919859/133053153-7080abd8-3d0c-42b0-8809-a781120ebd33.png)

and a bit similar to color picker which allow to specific value and label.

![image](https://user-images.githubusercontent.com/2919859/133053425-b8540daf-0f40-4676-88da-9d6315c40780.png)

**Comment by bjarnef** [comment]
Regarding color picker devlopers may just use the color code directly as inline styles in frontend, while in other scenarios it makes more sence the store a different value like a value used as class name or data attribute value. Currently we would have to either transform the color in view or controller, create our own color picker property editor or use a package like Contentment.

**Comment by SofusMunch** [comment]
I was just about to write my own editors to do this exact thing. 

I don’t think I ever use dropdowns, radio buttons or checkboxes without having to parse them in the view. More often than not I also have to code a fall back value if nothing was selected.

I would like be able to:
- Set label/values
- Set default value(s)
- Possibly set min/max for multiple dropdowns/checkboxes

**Comment by bjarnef** [comment]
@iOvergaard @nielslyngsoe in the new backoffice the multivalues property editors currently expect a string arrary, e.g. a config like this:

```
"Configuration": {
    "fileExtensions": [
      "mp3",
      "weba",
      "oga",
      "opus"
    ]
  }
 ```
 
it would be useful if these can support `label/value` as well as the value if often not want you want to be shown to content editors - especially when dealing with label translations in backoffice - and value is not always what one want to present to content editors or the other way around.

E.g. Approved Color has this configuration:

```
"Configuration": {
    "items": [
      {
        "value": "000000",
        "label": "Black"
      },
      {
        "value": "ffffff",
        "label": "White"
      },
      {
        "value": "ff0000",
        "label": "Red"
      }
    ],
    "useLabel": true
  },
 ```

**Comment by bjarnef** [comment]
I was thinking a multi values property editor / component will be useful in both core, but also packages or when extending backoffice UI, mainly label/value or key/value and the (approved) color picker may be able to extend this.

For our projects we also have a custom Button Picker property editor, which allow to set an (optional) icon, (optional) label and (required) value.

Perhaps it makes sense to have a flexible multi values property editor and pass in an array of objects (similar to settings property in configuration of a property editor).

**Scenario 1: Label/value for lists (dropdown, radiobutton, checkbox)**

```
[
   {
      "alias": "label",
      "label": "Label",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": false
   },
   {
      "alias": "value",
      "label": "Value",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": true,
      "unique": true  // or something else to indicate value only must exist once.
   }
]
``` 

**Scenario 2: Crop dimensions**

```
[
   {
      "alias": "alias",
      "label": "Alias",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": true,
      "unique": true
   },
   {
      "alias":"width",
      "label":"Width",
      "propertyEditorUiAlias":"Umb.PropertyEditorUi.Integer",
      "mandatory":false
   },
   {
      "alias": "height",
      "label": "Height",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.Integer",
      "mandatory": false
   }
]
```

**Scenario 3: Currencies**

```
[
   {
      "alias": "isoCode",
      "label": "ISO 4217 Code",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": true
   },
   {
      "alias":"price",
      "label":"Price",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.Decimal",
      "mandatory": false
   }
]
```

**Scenario 4: Button Picker**
Alignment: (left, center, right), Formatting (bold, italic, underline), Format (square, landscape, portrait), etc.

```
[
   {
      "alias": "icon",
      "label": "Icon",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.IconPicker",
      "mandatory": false
   },
   {
      "alias":"Label",
      "label":"Label",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": false
   },
   {
      "alias": "value",
      "label": "Value",
      "propertyEditorUiAlias": "Umb.PropertyEditorUi.TextBox",
      "mandatory": true,
      "unique": true
   }
]
```

It would allow to re-use much of the existing property editor UIs and pre-configure items in the content type property.

Of course it doesn't cover every use-case, but it allow to pre-configure items in a property editor setting property, where each object value not necessary is a textbox/string.

**Reply by bjarnef** [reply]
@iOvergaard @nielslyngsoe would something like this makes sense? It seems to be familiar with current settings properties in configuration of property editors and allow a more flexible configuration.

Of course it may not look great with 10+ properties horizontally, but useful when just 2-5 properties per item/row. For more complex configuration custom property editor UIs / components can be added.

It may even solve the issue defined repeatable content item for something else than textbox/string - e.g. just a single property "Label" and then the "repeatable content property editor UI can probably map each item to return `Array<T>` - e.g. `string[]` but the propertyEditorUiAlias may be `Umb.PropertyEditorUi.TextBox` or `Umb.PropertyEditorUi.EmailAddress` or something else.
https://github.com/umbraco/Umbraco-CMS/discussions/17566

**Reply by skttl** [reply]
Just want to put this one in here :)
https://github.com/leekelleher/umbraco-package-ideas/issues/47

Maybe also this one:
https://github.com/leekelleher/umbraco-package-ideas/issues/48

**Comment by bjarnef** [comment]
I have something like this custom property editor (a combination of (approved) color picker and the property editors dropdown, radiobutton list and checkbox list based on same multi value editor).

It would be useful it Umbraco core had a model with e.g. `label` and `value` properties, so I just can extend from this with an `icon` (or additional properties).

<img width="2246" height="714" alt="image" src="https://github.com/user-attachments/assets/866f65c2-4d86-4349-8d84-2b6d1be14a49" />

<img width="578" height="175" alt="image" src="https://github.com/user-attachments/assets/1d2c2e59-f13b-410d-87ae-7a9e4c0a537d" />

**Reply by jdpnielsen** [reply]
I have a custom version of this in all my projects. I think it should be included in core, as its generally useful for all projects, but also a very useful building block for package developers.

**Reply by skttl** [reply]
I usually use Contentment Data List for this with the buttons editor and a userdefined list

**Reply by jdpnielsen** [reply]
@skttl Do you build packages that use Contentment as a dependency, or how does that work?

**Reply by bjarnef** [reply]
@skttl I like Contentment package for larger projects or when used to a higher degree in project, but for mainly simple projects I prefer to used existing features in Umbraco and in this case combined with a simple property editor.
